sourcegraph/sourcegraph
# Remove ThemeProps and add new theme state logic

## Before

Before this PR we had theme state on the top level in the Layout.tsx, we had a hook that provided an object with public API for theme (to read or to set a new value), there were at least two problems 
- Complex implementation of useThemeProps (rxjs to listen browser theme, mixed with logic around setting theme CSS classes)
- Prop drilling, after we created these theme API on the top level, it was distributed via props

I think I don't need to dive deep in why prop drilling is a bad practice, but still, with prop drilling and nested components, we were updating the whole react tree when theme was changed, also passing these public Theme props required extending your component props with special type `ThemeProps` and this becomes a mess quickly if you extend something else (like one component props extended other component props)

## Now

In this PR, we use more standard ways for global state distirubiotion - Context. In the new `useTheme` hook, we don't do any effects with CSS classes (consumer on the top level should handle it), so it's safe to call useTheme on any UI level. Privosly it wasn't safe, and because of this, we had this problem https://github.com/sourcegraph/sourcegraph/issues/47501

## Further steps

Ideally we shouldn't have any JS logic about theme besides CSS classes set on the document element. And all colors should work through CSS variables *custom-properties, but at the moment, we have at least two major places where we have to have JS check about theme. 
- Brand logo (since we use svg as image, in img src attribute), we should change URL based on theme (in dark theme brand logo has difference color value). I'm wondering that we probably can change it and use svg directly without img tag)
- Monaco editor doesn't support colors through CSS variables and we change set of colors in js runtime based on theme value, I don't know maybe with CodeMirror we could use just CSS variables and remove this logic 


## Test plan
- Check web app with different theme values
- Check jetbrains plugin 
- Check vscode webview extension 
- Check browser extension UI 
- Check code host integration tooltip UI 



## App preview:

- [Web](https://sg-web-vk-improve-theme-state.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.
